src/main/scala/loamstream/compiler/LoamFile.scala

```diff
@@ -0,0 +1,40 @@
+package loamstream.compiler
+
+import loamstream.loam.LoamSyntax
+import loamstream.loam.LoamScriptContext
+import loamstream.loam.LoamProjectContext
+import loamstream.loam.LoamScript
+
+/**
+ * @author clint
+ * May 28, 2020
+ */
+abstract class LoamFile extends LoamSyntax with LoamScript.LoamScriptBox {
+  override def projectContext: LoamProjectContext = {
+    val context = LoamFile.ContextHolder.projectContext
+    
+    require(
+        context != null,
+        s"No ${LoamProjectContext.getClass.getSimpleName} set.  Set it with ContextHolder.projectContext = ...")
+    
+    context
+  }
+
+  override implicit lazy val scriptContext: LoamScriptContext = new LoamScriptContext(projectContext)
+}
+
+object LoamFile {
+  /**
+   * @author clint
+   * May 28, 2020
+   */
+  private[compiler] object ContextHolder {
+    private val contextVar: ThreadLocal[LoamProjectContext] = new ThreadLocal
```

Comment:
Obviously the code works as-is, but wanted to propose a possible alternative: `scala.util.DynamicVariable` is thread local as well, and scoped (dynamically). Usually when I see thread local variables w/ a "clear" option, it's a bit of a code smell that says "this should only be set during X". Instead, you could do:

```scala
private val context = new DynamicVariable[LoamProjectContext](null)

def withContext[A](ctx: LoamProjectContext)(body: => A): A = {
  context.withValue(ctx)(body)
}
```

Now, in `body` the context is valid, and not outside of body.

Reply:
Thanks, I didn't know this class existed before.  It looks like it provides functionality similar to what I've added here with `setProjectContextAndThen`, `doInNewThread`, etc, but obviously a standard-lib solution is preferable to something hand-rolled.

With `DynamicVariable`, I'd still need `doInNewThread` though, right, to ensure that each compilation-evaluation run gets its own unique `LoamProjectContext`?

Reply:
I don't have the full context of the problem you're solving. It's possible it could be done w/ a simple implicit, too. Something like:

```scala
implicit val context = new LoamProjectContext(...)

// instantiate LoamFile, which has (implicit val context: LoamProjectContext)
// now the LoamFile calls -- and any functions they call that need it -- have a context
```

Potentially a big change, though. `DynamicVariable` is just an alternative to an implicit that's a bit more in line with what you were doing before.